### DISCUSSION TITLE
Can't build xmake on WSL using mingw (building for Windows)

### DISCUSSION THREAD

**Discussion by AndreHunf** [xmake-io/xmake#5821] [Q&A]
I've recently been trying to compile xmake for Windows using WSL along with mingw. According to the documentation, I'm following the command below to compile the source code:

`xmake f -p mingw -a x86_64`

When using the configuration above, I encounter the following warning message:

```
warning: add_cxflags("-Wno-error=nullability-completeness") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
warning: add -v for getting more warnings ..
```

However, at a given moment, when typing `xmake` to build the code, the compilation stops showing the following error:

```
error: src/tbox/tbox/src/tbox/platform/windows/file.c:446:109: error: ‘COPY_FILE_COPY_SYMLINK’ undeclared (first use in this function); did you mean ‘TB_FILE_COPY_LINK’?
  446 |         if (tb_kernel32()->CopyFileExW && tb_kernel32()->CopyFileExW(full0, full1, tb_null, tb_null, FALSE, COPY_FILE_COPY_SYMLINK))
      |                                                                                                             ^~~~~~~~~~~~~~~~~~~~~~
      |                                                                                                             TB_FILE_COPY_LINK
src/tbox/tbox/src/tbox/platform/windows/file.c:446:109: note: each undeclared identifier is reported only once for each function it appears in
  > in src/tbox/tbox/src/tbox/platform/file.c
warning: add_cxflags("-Wno-error=nullability-completeness") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
warning: add -v for getting more warnings ..
```

I tried every possible configuration to get rid of this error, but with no success. 

So the question is: what steps should I do to compile the xmake for Windows using WSL and mingw?

Thanks in advance.

**Comment by star-hengxing** [comment]
https://github.com/mingw-w64/mingw-w64/blob/cf5c50cce0b4f8610eafd95443d1d3767de386a5/mingw-w64-headers/include/winbase.h#L130

Maybe missing this code. You can open a issue
```c
#if defined(__MINGW32__) || defined(__MINGW64__)
#   undef _WIN32_WINNT
#   define _WIN32_WINNT 0x0600
#endif
```

**Reply by waruqi** [reply]
try it again. please re-clone dev branch.

**Reply by AndreHunf** [reply]
The error seems to disappear, but now at 90% of the compilation I get the following error:

```
error: src/xmake/hash/xxhash.c:86:76: error: ‘buffer’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
   86 |         for (i = 0; i < n; ++i) tb_snprintf(s + (i << 1), 3, "%02x", buffer[i]);
      |                                                                      ~~~~~~^~~
cc1: all warnings being treated as errors
```

With this warning at the end:

```
  > in src/xmake/hash/xxhash.c
warning: add_cxflags("-Wno-error=nullability-completeness") is ignored, please pass `{force = true}` or call `set_policy("check.auto_ignore_flags", false)` if you want to set it.
warning: add -v for getting more warnings ..
```

**Reply by waruqi** [reply]
try it again

**Reply by AndreHunf** [reply]
The compilation worked! Aside from some warnings at 67% of compilation like above:

```
src/lua-cjson/lua-cjson/dtoa.c: In function ‘dtoa’:
src/lua-cjson/lua-cjson/dtoa.c:3974:27: warning: ‘j1’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 3974 |     dval(&eps1) *= tens[j1 & 0xf];
      |                        ~~~^~~~~
src/lua-cjson/lua-cjson/dtoa.c:3691:6: note: ‘j1’ was declared here
 3691 |   j, j1, k, k0, k_check, leftright, m2, m5, s2, s5,
      |      ^~
[ 67%]: cache compiling.release src/xmake/lz4/decompress_stream_open.c
src/lua-cjson/lua-cjson/strbuf.c: In function ‘debug_stats’:
src/lua-cjson/lua-cjson/strbuf.c:98:17: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   98 |                 (long)s, s->reallocs, s->length, s->size);
      |                 ^
src/lua-cjson/lua-cjson/strbuf.c: In function ‘strbuf_resize’:
src/lua-cjson/lua-cjson/strbuf.c:172:17: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  172 |                 (long)s, s->size, newsize);
      |                 ^
```

I finally got the xmake.exe, however, when I try to run on Windows from cmd (i.e. `xmake --help`), nothing happens. I took all the /build folder outside WSL to my Windows desktop.

**Reply by waruqi** [reply]
> Aside from some warnings at 67% of compilation like above:

it's warnings from the third-party libs, you can open issues for these libs.

>  however, when I try to run on Windows from cmd (i.e. xmake --help), nothing happens. I took all the /build folder outside WSL to my Windows desktop.

you need set XMAKE_PROGRAM_DIR -> lua scripts

see https://github.com/xmake-io/xmake/blob/ad8324471f281a348a01fd982d483bb9cf0e56cf/scripts/srcenv.profile#L2

**Reply by AndreHunf** [reply]
My apologies, but I didn't understand the last part. Could you guide me?

**Reply by waruqi** [reply]
you need set `XMAKE_PROGRAM_DIR ` environment var to xmake script directory.  https://github.com/xmake-io/xmake/tree/dev/xmake

**Reply by AndreHunf** [reply]
When running the script I get this error:

`error: cannot open C:\Users\Avell\Softwares\xmake\scripts\..\xmake/core/_xmake_main.lua: No such file or directory`

**Reply by waruqi** [reply]
you need set correct XMAKE_PROGRAM_DIR

**Reply by AndreHunf** [reply]
I did everything as you instructed. I set the XMAKE_PROGRAM_DIR to the specified [directory](https://github.com/xmake-io/xmake/tree/dev/xmake) as pointed out and also set the XMAKE_PROGRAM_FILE to the path of the generated `xmake.exe`. I even edited the `srcenv.bat` switching the backslash to the forward slash thinking that could be the problem, but nothing worked.

I also downloaded and tried to run `xmake-v2.9.6.win64` from the releases page with another machine and I got the same error as my compiled version.

At this point I could not trace where this error is coming from. I might open an issue relating this problem, if you prefer.

**Reply by waruqi** [reply]
> C:\Users\Avell\Softwares\xmake\scripts\..\xmake/core/_xmake_main.lua: No such file or directory

you can check this file path. 

https://github.com/xmake-io/xmake/blob/dev/xmake/core/_xmake_main.lua

**Reply by AndreHunf** [reply]
Sorry but, if even the compiled version found in the releases page shows the same error I encounter, there must be a bug of some sort. I'll be back when a solution for this problem is found.

**Reply by waruqi** [reply]
> > C:\Users\Avell\Softwares\xmake\scripts..\xmake/core/_xmake_main.lua: No such file or directory
> 
> you can check this file path.
> 
You didn't give me any feedback info, so I can't confirm what problem you encountered. Does the file path exist?

**Reply by AndreHunf** [reply]
> Does the file path exist?

The path does exist and there is the `_xmake_main.lua` file, but somehow xmake cannot find this file/directory and I cannot trace from which step this error is comming from.

> You didn't give me any feedback info, so I can't confirm what problem you encountered.

Is there any kind of information you are looking for? I'm willing to provide the necessary information to resolve this issue.

Below I will show the steps I took until I reached the current error:
1. Download xmake with `git clone --recursive https://github.com/xmake-io/xmake.git`;
2. Navigate to `/core` and configure xmake with `xmake f -p mingw -a x86_64`;
3. After configuration, build the source code with `xmake` command;
4. After finishing the compilation, bring all xmake folder from WSL to Windows;
5. On Windows, navigate to `/scripts` and execute `srcenv.bat` from cmd. Windows will open another cmd with the xmake root directory (with environment vars set);
6. Type `xmake --version` to check if everything is working right;
7. Cmd shows `error: cannot open C:\Users\Avell\Softwares\xmake\scripts\..\xmake/core/_xmake_main.lua: No such file or directory`

If you need more information, I am open to assist you.
